loki/bulk/item.py

```diff
@@ -118,691 +154,1278 @@ def __eq__(self, other):
-    def path(self):
+    def concretize_dependencies(self):
         """
-        The filepath of the associated source file
-        """
-        return self.source.path
+        Trigger a re-parse of the source file corresponding to the current item's scope
 
-    @property
-    def role(self):
+        This uses :attr:`_depends_class` to determine all :any:`RegexParserClass` that
+        the are require to be parsed to find the item's dependencies.
         """
-        Role in the transformation chain, for example ``'driver'`` or ``'kernel'``
-        """
-        return self.config.get('role', None)
+        if not self._depends_class:
+            return
+        scope = self.scope_ir
+        while scope.parent:
+            scope = scope.parent
+        if hasattr(scope, 'make_complete'):
+            scope.make_complete(frontend=REGEX, parser_classes=self._depends_class)
 
-    @property
-    def mode(self):
+    def create_definition_items(self, item_factory, config=None, only=None):
         """
-        Transformation "mode" to pass to the transformation
+        Create the :any:`Item` nodes corresponding to the definitions in the
+        current item
+
+        Parameters
+        ----------
+        item_factory : :any:`ItemFactory`
+            The :any:`ItemFactory` to use when creating the items
+        config : :any:`SchedulerConfig`, optional
+            The scheduler config to use when instantiating new items
+        only : list of :any:`Item` classes
+            Filter the generated items to include only those provided in the list
+
+        Returns
+        -------
+        tuple
+            The list of :any:`Item` nodes
         """
-        return self.config.get('mode', None)
+        items = as_tuple(flatten(
+            item_factory.create_from_ir(node, self.scope_ir, config)
+            for node in self.definitions
+        ))
+        items = as_tuple(item for item in items if item is not None)
+        if only:
+            items = tuple(item for item in items if isinstance(item, only))
+        return items
 
-    @property
-    def expand(self):
+    def create_dependency_items(self, item_factory, config=None, only=None):
         """
-        Flag to trigger expansion of children under this node
+        Create the :any:`Item` nodes corresponding to the dependencies of the
+        current item
+
+        Parameters
+        ----------
+        item_factory : :any:`ItemFactory`
+            The :any:`ItemFactory` to use when creating the items
+        config : :any:`SchedulerConfig`, optional
+            The scheduler config to use when instantiating new items
+        only : list of :any:`Item` classes
+            Filter the generated items to include only those provided in the list
+
+        Returns
+        -------
+        tuple
+            The list of :any:`Item` nodes
         """
-        return self.config.get('expand', False)
+        if not (dependencies := self.dependencies):
```

Comment:
So, I'm not 100% that I'm understanding this right, but to illustrate my point on expensive SGraph rebuilds, I believe this lines is triggered for determining every edge in the base graph, and will cause a regex-parse of every dependency item (via `concretize_dependencies`), regardless of whether this has been done before. I've not fully confirmed this, other than by observing that this call can take up to 2s for certain items. 

If this is the case, I'm not super-sure how to fix this, but it feels like this should be fixable.

Reply:
Ok for now, follow-up tracked via #254 .